bluemira/equilibria/opt_problems.py

```diff
@@ -0,0 +1,811 @@
+
+        # Set up bounds
+        bounds = self.get_mapped_state_bounds(self.region_mapper, max_currents)
+        # Add bounds information to help automatic differentiation of objective
+        self._objective._args["ad_args"] = {"bounds": bounds}
+        self._objective._args["objective_args"] = {
+            "coilset": coilset,
+            "eq": eq,
+            "targets": targets,
+            "region_mapper": self.region_mapper,
+            "current_scale": self.scale,
+            "gamma": gamma,
+        }
+
+        # Set up optimiser
+        dimension = len(bounds[0])
+        self.set_up_optimiser(dimension, bounds)
+
+    def get_mapped_state_bounds(self, region_mapper: RegionMapper, max_currents):
+        """
+        Get mapped bounds on the coilset state vector from the coil regions and
+        maximum coil currents.
+
+        Parameters
+        ----------
+        region_mapper: RegionMapper
+            RegionMapper mapping coil positions within the allowed optimisation
+            regions.
+        max_currents float or np.array(len(coilset._ccoils)) (default = None)
+            Maximum allowed current for each independent coil current in coilset [A].
+            If specified as a float, the float will set the maximum allowed current
+            for all coils.
+
+        Returns
+        -------
+        bounds: np.array
+            Array containing state vectors representing lower and upper bounds
+            for coilset state degrees of freedom.
+        """
+        # Get mapped position bounds from RegionMapper
+        _, lower_lmap_bounds, upper_lmap_bounds = region_mapper.get_Lmap(self.coilset)
+        current_bounds = self.get_current_bounds(self.coilset, max_currents, self.scale)
+
+        lower_bounds = np.concatenate((lower_lmap_bounds, current_bounds[0]))
+        upper_bounds = np.concatenate((upper_lmap_bounds, current_bounds[1]))
+        bounds = (lower_bounds, upper_bounds)
+        return bounds
+
+    def optimise(self):
+        """
+        Optimiser handle. Used in __call__
+
+        Returns
+        -------
+        self.coilset: CoilSet
+            Optimised CoilSet object.
+        """
+        # Get initial state and apply region mapping to coil positions.
+        initial_state, _ = self.read_coilset_state(self.coilset, self.scale)
+        _, _, initial_currents = np.array_split(initial_state, self.substates)
+        initial_mapped_positions, _, _ = self.region_mapper.get_Lmap(self.coilset)
+        initial_mapped_state = np.concatenate(
+            (initial_mapped_positions, initial_currents)
+        )
+
+        # Optimise
+        state = self.opt.optimise(initial_mapped_state)
+
+        # Call objective function final time on optimised state
+        # to set coilset.
+        # Necessary as optimised state may not always be the final
+        # one evaluated by optimiser.
+        self.get_state_figure_of_merit(state)
+        return self.coilset
+
+    @staticmethod
```

Comment:
I take it there is a good reason for using a static method and dodging a super call? What is it?

Reply:
Primarily to make re-use of objectives easier, from other OptimisationProblems - if the methods are not static, NLOpt will complain if the function is used elsewhere as the first argument will no longer be the state vector.

This way, moving objective functions to the opt_objectives library, or calling them in another optimisation problem, doesn't require any changes to the function.